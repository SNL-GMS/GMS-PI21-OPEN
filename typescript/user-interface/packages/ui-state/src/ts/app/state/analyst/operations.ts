import type { LegacyEventTypes } from '@gms/common-model';
import { UILogger } from '@gms/ui-util';

import { selectPreferredFilterListForActivity } from '../../api/signal-enhancement-configuration/selectors';
import { signalEnhancementConfigurationApiSlice } from '../../api/signal-enhancement-configuration/signal-enhancement-api-slice';
import type { AppDispatch, AppState } from '../../store';
import { analystActions } from './analyst-slice';
import { selectSelectedFilterList, selectSelectedFilterListName } from './selectors';
import { WaveformDisplayMode, WaveformSortType } from './types';

const logger = UILogger.create('[UI State Manager - Analyst]', process.env.GMS_LOG_UI_STATE_STORE);

/**
 * Redux operation for setting the mode.
 *
 * @param mode the mode to set
 */
export const setMode = (mode: WaveformDisplayMode) => (dispatch: AppDispatch): void => {
  dispatch(analystActions.setMode(mode));
};

/**
 * Redux operation for setting the measurement mode entries.
 *
 * @param entries the measurement mode entries to set
 */
export const setMeasurementModeEntries = (entries: Record<string, boolean>) => (
  dispatch: AppDispatch
): void => {
  dispatch(analystActions.setMeasurementModeEntries(entries));
};

/**
 * Redux operation for setting the selected location solution.
 *
 * @param locationSolutionSetId the location solution set id
 * @param locationSolutionId the location solution id
 */
export const setSelectedLocationSolution = (
  locationSolutionSetId: string,
  locationSolutionId: string
) => (dispatch: AppDispatch, getState: () => AppState): void => {
  if (getState().app.analyst.location.selectedLocationSolutionSetId !== locationSolutionSetId) {
    dispatch(analystActions.setSelectedLocationSolutionSetId(locationSolutionSetId));
  }

  if (getState().app.analyst.location.selectedLocationSolutionId !== locationSolutionId) {
    dispatch(analystActions.setSelectedLocationSolutionId(locationSolutionId));
  }
};

/**
 * Redux operation for setting the selected preferred location solution.
 *
 * @param preferredLocationSolutionSetId the preferred location solution set id
 * @param preferredLocationSolutionId the preferred location solution id
 */
export const setSelectedPreferredLocationSolution = (
  preferredLocationSolutionSetId: string,
  preferredLocationSolutionId: string
) => (dispatch: AppDispatch, getState: () => AppState): void => {
  if (
    getState().app.analyst.location.selectedPreferredLocationSolutionSetId !==
    preferredLocationSolutionSetId
  ) {
    dispatch(
      analystActions.setSelectedPreferredLocationSolutionSetId(preferredLocationSolutionSetId)
    );
  }

  if (
    getState().app.analyst.location.selectedPreferredLocationSolutionId !==
    preferredLocationSolutionId
  ) {
    dispatch(analystActions.setSelectedPreferredLocationSolutionId(preferredLocationSolutionId));
  }
};

/**
 * Redux operation for setting the current open event id.
 *
 * @param event the event to set
 * @param latestLocationSolutionSet
 * @param preferredLocationSolutionId
 */
export const setOpenEventId = (
  event: LegacyEventTypes.Event | undefined,
  latestLocationSolutionSet: LegacyEventTypes.LocationSolutionSet | undefined,
  preferredLocationSolutionId: string | undefined
) => (dispatch: AppDispatch, getState: () => AppState): void => {
  if (getState().app.workflow.timeRange && event) {
    if (getState().app.analyst.openEventId !== event.id) {
      dispatch(analystActions.setOpenEventId(event.id));
      dispatch(analystActions.setSelectedEventIds([event.id]));
      dispatch(analystActions.setSelectedSortType(WaveformSortType.distance));

      // set the default (latest) location solution
      setSelectedLocationSolution(
        latestLocationSolutionSet ? latestLocationSolutionSet.id : undefined,
        latestLocationSolutionSet ? latestLocationSolutionSet.locationSolutions[0].id : undefined
      )(dispatch, getState);

      // set the default (latest) preferred location solution
      setSelectedPreferredLocationSolution(
        latestLocationSolutionSet.id,
        preferredLocationSolutionId
      )(dispatch, getState);
    }
  } else {
    dispatch(analystActions.setOpenEventId(undefined));
    dispatch(analystActions.setSelectedEventIds([]));
    dispatch(analystActions.setSelectedSortType(WaveformSortType.stationNameAZ));
    dispatch(analystActions.setMeasurementModeEntries({}));
    // update the selected location and preferred location solutions
    setSelectedLocationSolution(undefined, undefined)(dispatch, getState);
    setSelectedPreferredLocationSolution(undefined, undefined)(dispatch, getState);
  }

  // TODO: confirm should we be clearing????
  dispatch(analystActions.setSelectedSdIds([]));
  dispatch(analystActions.setSdIdsToShowFk([]));
  setMode(WaveformDisplayMode.DEFAULT)(dispatch);
};

/**
 * creates an action that sets the selected filter list to be the preferred filter list from the list of
 * preferred filter lists returned by the filter definition query.
 * Logs warnings if it is being misused, for example, if the preferred filter list for the activity is nullish,
 * or if the selectedFilterList was already set.
 *
 * @see signalEnhancementConfigurationApiSlice
 */
export const setPreferredFilterList = () => (
  dispatch: AppDispatch,
  getState: () => AppState
): void => {
  const state = getState();
  const preferredFilterListForActivity = selectPreferredFilterListForActivity(state);
  const selectedFilterList = selectSelectedFilterListName(state);
  if (selectedFilterList == null && preferredFilterListForActivity != null) {
    dispatch(analystActions.setSelectedFilterList(preferredFilterListForActivity));
  } else if (selectedFilterList != null && preferredFilterListForActivity == null) {
    logger.warn(`Cannot set preferred filter list because it is not defined.`);
  } else if (selectedFilterList == null && preferredFilterListForActivity == null) {
    logger.warn(
      `Cannot set preferred filter list because neither it nor the selected filter list is defined.`
    );
  } else {
    logger.warn(
      `Cannot set preferred filter list ${preferredFilterListForActivity} because ${selectedFilterList} was already selected.`
    );
  }
};

/**
 * Sets the default filter from the currently selected filter list.
 * Logs a warning if no selected filter is found.
 * This will subscribe the component to the results of the filter list query.
 * TODO: We need to unsubscribe? How? Where?
 */
export const setDefaultFilter = () => async (
  dispatch: AppDispatch,
  getState: () => AppState
): Promise<void> => {
  const state = getState();
  // subscribe to this state
  await dispatch(
    signalEnhancementConfigurationApiSlice.endpoints.getFilterListsDefinition.initiate()
  );
  const fl = selectSelectedFilterList(state);
  if (fl == null) {
    logger.warn('cannot set default filter list');
    return;
  }
  dispatch(analystActions.setSelectedFilterIndex(fl.defaultFilterIndex));
};
