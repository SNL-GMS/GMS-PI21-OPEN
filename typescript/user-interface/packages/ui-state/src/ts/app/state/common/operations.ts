import { Displays } from '@gms/common-model';
import { areListsSame, Timer } from '@gms/common-util';
import produce from 'immer';
import { batch } from 'react-redux';

import type { AppDispatch, AppState } from '../../store';
import { commonSlice } from './common-slice';
import { isKeyboardShortcutPopupOpen } from './selectors';
import type { CommonState, GLDisplayState } from './types';

// return true if the visibility has changed, false otherwise
const commandPaletteVisibilityHasChanged = (visibility: boolean, currentVisibility: boolean) =>
  visibility !== currentVisibility;

/**
 * Redux operation to set the visibility of the command palette
 *
 * @param visibility the true/false state of whether or not the command palette should be visible
 */
export const setCommandPaletteVisibility = (visibility: boolean) => (
  dispatch: AppDispatch,
  getState: () => AppState
): void => {
  const state = getState();
  const { common } = state.app;
  if (commandPaletteVisibilityHasChanged(visibility, common.commandPaletteIsVisible)) {
    if (isKeyboardShortcutPopupOpen(state) && visibility) {
      dispatch(commonSlice.actions.setKeyboardShortcutsVisibility(false));
    }
    dispatch(commonSlice.actions.setCommandPaletteVisibility(visibility));
  }
};

/**
 * @param ids new set of selected ids
 * @returns whether they are the same, irrespective of order
 */
export const idsHaveChanged = (ids: string[]): ((getState: () => AppState) => boolean) => (
  getState: () => AppState
): boolean => {
  Timer.start('[common workspace operations] idsHaveChanged');
  // Sorting to ensure actual different selections ex: [2, 1] [1, 2] are not really different
  const state: CommonState = getState().app.common;
  const equal = !areListsSame(ids, state.selectedStationIds);
  Timer.end('[common workspace operations] idsHaveChanged');
  return equal;
};

/**
 * Overwrites the selected stations in the Redux state.
 *
 * @param ids the ids to set
 */
export const setSelectedStationIds = (ids: string[]) => (
  dispatch: AppDispatch,
  getState: () => AppState
): void => {
  if (idsHaveChanged(ids)(getState)) {
    dispatch(commonSlice.actions.setSelectedStationIds(ids));
  }
};

/**
 * Sets a single display's state in the glLayoutState tracked in Redux.
 *
 * @param displayName a string uniquely identifying a golden layout display
 * @param displayState the state of the display (open, closed, etc)
 * @returns a dispatch function that dispatches the state update.
 */
export const setGlDisplayState = (displayName: string, displayState: GLDisplayState) => (
  dispatch: AppDispatch,
  getState: () => AppState
): void => {
  if (displayName && displayState && Displays.isValidDisplayName(displayName)) {
    batch(() => {
      const { glLayoutState } = getState().app.common;
      const newLayoutState = produce(glLayoutState, draft => {
        draft[displayName] = displayState;
      });
      dispatch(commonSlice.actions.setGlLayoutState(newLayoutState));
    });
  } else if (displayName && !Displays.isValidDisplayName(displayName)) {
    throw new Error(`Invalid display name: ${displayName} is not a name of a known display.`);
  }
};
