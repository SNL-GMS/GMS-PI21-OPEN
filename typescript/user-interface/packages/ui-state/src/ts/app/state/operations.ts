import isEqual from 'lodash/isEqual';
import { batch } from 'react-redux';

import type { AppDispatch, AppState } from '../store';
import * as Actions from './actions';
import type { AuthenticationStatus } from './user-session/types';
import { userSessionSlice } from './user-session/user-session-slice';

/**
 * Redux operation for setting the authentication status.
 *
 * @param event the event to set
 * @param authStatus
 */
export const setAppAuthenticationStatus = (authStatus: AuthenticationStatus) => (
  dispatch: AppDispatch,
  getState: () => AppState
): void => {
  const state: AppState = getState();
  if (!isEqual(state.app.userSession.authenticationStatus, authStatus)) {
    if (!authStatus.userName && !authStatus.authenticated) {
      batch(() => {
        // reset the application state
        dispatch(Actions.reset());

        // update the authentication status
        dispatch(userSessionSlice.actions.setAuthenticationStatus(authStatus));
      });
    } else {
      batch(() => {
        // update the authentication status
        dispatch(userSessionSlice.actions.setAuthenticationStatus(authStatus));
      });
    }
  }
};
