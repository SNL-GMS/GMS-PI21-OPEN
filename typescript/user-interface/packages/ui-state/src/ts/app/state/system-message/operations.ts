import type { SystemMessageTypes } from '@gms/common-model';
import orderBy from 'lodash/orderBy';
import reverse from 'lodash/reverse';
import { batch } from 'react-redux';

import type { AppDispatch, AppState } from '../../store';
import { systemMessageSlice } from './system-message-slice';
import type { SystemMessageState } from './types';

/**
 * Redux operation that adds system messages to the state.
 *
 * @param messages the system messages to add
 * @param limit (optional) limit the number of messages in the redux state
 * when adding new messages; if set when adding new messages the message list
 * result will not be larger than the size of the `limit` specified.
 * @param pageSizeBuffer (optional) the size of the page buffer; if specified and the
 * limit is reached then messages will be removed at increments of the page buffer size
 */
export const addSystemMessages = (
  messages: SystemMessageTypes.SystemMessage[],
  limit?: number /* default to delete zero messages */,
  pageSizeBuffer?: number
) => (dispatch: AppDispatch, getState: () => AppState): void => {
  const state: SystemMessageState = getState().app.systemMessage;
  if (messages && messages.length > 0) {
    // batch the dispatches - this will only result in one combined re-render, not two
    batch(() => {
      const allMessages = state.systemMessages
        ? [...state.systemMessages, ...messages]
        : [...messages];

      // apply the size limit if necessary before updating the redux state
      const actualLimit = limit && limit > 0 ? limit : 0;
      if (actualLimit > 0 && allMessages.length > actualLimit) {
        const actualPageSizeBuffer =
          pageSizeBuffer && pageSizeBuffer > 0 && pageSizeBuffer < limit ? pageSizeBuffer : 0;
        if (actualPageSizeBuffer > 0) {
          // remove page size increments until the number of messages is less than the limit
          let updatedMessages = reverse([...allMessages]);
          while (updatedMessages.length > limit) {
            updatedMessages = updatedMessages.splice(
              0,
              updatedMessages.length - actualPageSizeBuffer
            );
          }
          dispatch(systemMessageSlice.actions.setSystemMessages(reverse([...updatedMessages])));
        } else {
          // ensure that the list size is equal to the `limit`
          dispatch(
            systemMessageSlice.actions.setSystemMessages(
              reverse(reverse(allMessages).splice(0, limit))
            )
          );
        }
      } else {
        dispatch(systemMessageSlice.actions.setSystemMessages(allMessages));
      }
      dispatch(
        systemMessageSlice.actions.setLatestSystemMessages(
          orderBy<SystemMessageTypes.SystemMessage>(messages, ['time'], ['desc'])
        )
      );
      dispatch(systemMessageSlice.actions.setLastUpdated(Date.now()));
    });
  }
};

/**
 * Redux operation that clears all system messages to the state.
 */
export const clearAllSystemMessages = () => (dispatch: AppDispatch): void => {
  // batch the dispatches - this will only result in one combined re-render, not two
  batch(() => {
    dispatch(systemMessageSlice.actions.setSystemMessages([]));
  });
};
