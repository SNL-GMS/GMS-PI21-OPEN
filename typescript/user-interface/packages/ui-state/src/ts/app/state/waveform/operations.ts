import type { ChannelTypes, CommonTypes, StationTypes, WorkflowTypes } from '@gms/common-model';
import { WaveformTypes } from '@gms/common-model';
import { UILogger } from '@gms/ui-util';
import produce from 'immer';
import flatMap from 'lodash/flatMap';
import isEqual from 'lodash/isEqual';
import sortBy from 'lodash/sortBy';
import uniq from 'lodash/uniq';
import { batch } from 'react-redux';

import { processingConfigurationApiSlice } from '../../api/processing-configuration/processing-configuration-api-slice';
import type { StationGroupsByNamesProps } from '../../api/station-definition/station-definition-api-slice';
import { stationDefinitionSlice } from '../../api/station-definition/station-definition-api-slice';
import { UIStateError } from '../../error-handling/ui-state-error';
import type { AppDispatch, AppState } from '../../store';
import type {
  StationVisibilityChanges,
  StationVisibilityChangesDictionary,
  WaveformState,
  ZoomIntervalProperties
} from './types';
import { ZOOM_INTERVAL_TOO_LARGE_ERROR_MESSAGE } from './types';
import * as Utils from './util';
import { waveformActions } from './waveform-slice';

const logger = UILogger.create('[UI State Manager - Waveform]', process.env.GMS_LOG_UI_STATE_STORE);

/**
 * Checks for null and undefined of the replacement station visibility map
 *
 * @param sVis
 * @returns boolean false if input is undefined or null
 */
export const stationsVisibilityDictionaryIsDefined = (
  sVis: StationVisibilityChangesDictionary
): boolean => {
  return typeof sVis !== 'undefined' && sVis !== null;
};

/**
 * Overwrites the StationsVisibility dictionary in the Redux state.
 *
 * @param sVis a map from station names (string) to StationVisibilityObjects
 */
export const setStationsVisibility = (sVis: StationVisibilityChangesDictionary) => (
  dispatch: AppDispatch
): void => {
  if (stationsVisibilityDictionaryIsDefined(sVis)) {
    dispatch(waveformActions.setStationsVisibility(sVis));
  }
};

/**
 * Internal function for resetting the station visibility to an empty map.
 *
 * @param dispatch the redux dispatch function
 */
export const resetStationsVisibility = (dispatch: AppDispatch): void => {
  dispatch(waveformActions.setStationsVisibility({}));
};

/**
 * Internal function for resetting the viewable interval and zoom interval to their default state: null.
 *
 * @param dispatch the redux dispatch function
 */
export const resetWaveformIntervals = (dispatch: AppDispatch): void => {
  batch(() => {
    dispatch(waveformActions.setViewableInterval(null));
    dispatch(waveformActions.setZoomInterval(null));
  });
};

/**
 * Internal function to calculate the boundaries for zoomInterval
 *
 * @param waveform
 */
export const calculateZoomIntervalProperties = (
  waveform: WaveformState
): ZoomIntervalProperties => {
  const viewableInterval = waveform.viewableInterval ?? null;
  const minOffset = waveform.minimumOffset ?? 0;
  const maxOffset = waveform.maximumOffset ?? 0;
  const startTimeSecs = viewableInterval?.startTimeSecs ?? 0;
  const endTimeSecs = viewableInterval?.endTimeSecs ?? 0;
  let diffStartTimeSecs = startTimeSecs + minOffset;
  let diffEndTimeSecs = endTimeSecs + maxOffset;
  let startTimeDiff = 0;
  let endTimeDiff = 0;
  const baseStationTime = waveform.baseStationTime ?? null;
  const prevZoomInterval = waveform.zoomInterval ?? null;
  if (baseStationTime) {
    startTimeDiff = baseStationTime - diffStartTimeSecs;
    endTimeDiff = diffEndTimeSecs - baseStationTime;
    if (startTimeDiff < endTimeDiff) {
      diffStartTimeSecs -= endTimeDiff - startTimeDiff;
      endTimeDiff = 0;
    } else {
      diffEndTimeSecs += startTimeDiff - endTimeDiff;
      startTimeDiff = 0;
    }
  }
  return {
    prevZoomInterval,
    viewableInterval,
    diffStartTimeSecs,
    diffEndTimeSecs,
    minOffset,
    maxOffset,
    startTimeDiff,
    endTimeDiff
  };
};

/**
 * Sets the zoom interval to the range provided. Rounds to the nearest millisecond.
 *
 * @param zoomInterval the interval to set it to. Must be within the viewable interval, and non-nullish.
 * @throws if the zoom interval is outside of the viewable interval, or if the viewable
 * interval is not set.
 */
export const setZoomInterval = (zoomInterval: CommonTypes.TimeRange) => (
  dispatch: AppDispatch,
  getState: () => AppState
): void => {
  const { waveform } = getState().app;
  const {
    prevZoomInterval,
    viewableInterval,
    diffStartTimeSecs,
    diffEndTimeSecs,
    minOffset,
    maxOffset,
    startTimeDiff,
    endTimeDiff
  } = calculateZoomIntervalProperties(waveform);

  if (isEqual(zoomInterval, prevZoomInterval)) {
    return;
  }
  if (zoomInterval == null) {
    dispatch(waveformActions.setZoomInterval(null));
    return;
  }
  if (!viewableInterval) {
    throw new UIStateError(ZOOM_INTERVAL_TOO_LARGE_ERROR_MESSAGE);
  }

  let zoomIntervalInRange = zoomInterval;
  if (
    zoomInterval.startTimeSecs < diffStartTimeSecs ||
    zoomInterval.endTimeSecs > diffEndTimeSecs
  ) {
    logger.warn(
      `Cannot set a zoom interval that is outside of the viewable interval. zoomInterval: ${JSON.stringify(
        zoomInterval
      )}, viewableInterval: ${JSON.stringify(
        viewableInterval
      )}, offsets: ${minOffset}, ${maxOffset}, timeDiffs: ${startTimeDiff}, ${endTimeDiff}
      }. Truncating to fit within viewable interval.`
    );
    zoomIntervalInRange = {
      startTimeSecs: Math.max(
        zoomInterval.startTimeSecs,
        viewableInterval.startTimeSecs + minOffset
      ),
      endTimeSecs: Math.min(zoomInterval.endTimeSecs, viewableInterval.endTimeSecs + maxOffset)
    };
  }
  dispatch(waveformActions.setZoomInterval(zoomIntervalInRange));
};

/**
 * Initializes the viewable range and zoom interval based on the value of the current interval time range in
 * redux, and the lead and lag times set in processing configuration.
 */
export const initializeWaveformIntervals = () => (
  dispatch: AppDispatch,
  getState: () => AppState
): void => {
  const { timeRange: currentInterval } = getState().app.workflow;

  const processingAnalystConfigurationQuery = processingConfigurationApiSlice.endpoints.getProcessingAnalystConfiguration.select()(
    getState()
  );

  const leadBufferDuration = processingAnalystConfigurationQuery?.data?.leadBufferDuration ?? 0;
  const lagBufferDuration = processingAnalystConfigurationQuery?.data?.lagBufferDuration ?? 0;

  if (currentInterval.startTimeSecs != null && currentInterval.endTimeSecs != null) {
    const startTimeSecs = currentInterval.startTimeSecs - Number(leadBufferDuration);
    const endTimeSecs = currentInterval.endTimeSecs + Number(lagBufferDuration);

    const timeRange: CommonTypes.TimeRange = {
      startTimeSecs: Number.isNaN(startTimeSecs) ? undefined : startTimeSecs,
      endTimeSecs: Number.isNaN(endTimeSecs) ? undefined : endTimeSecs
    };
    batch(() => {
      dispatch(waveformActions.setViewableInterval(timeRange));
      setZoomInterval(timeRange)(dispatch, getState);
    });
  } else {
    throw new Error('Cannot initialize waveform intervals if no current interval is defined.');
  }
};

/**
 * Initializes the station visibility map based on the value of the current station group in redux
 */
export const initializeStationVisibility = (
  stationGroup: WorkflowTypes.StationGroup,
  effectiveTime: number
) => async (dispatch: AppDispatch, getState: () => AppState): Promise<void> => {
  const stationGroupQueryProps: StationGroupsByNamesProps = {
    stationGroupNames: [stationGroup.name],
    effectiveTime
  };

  await dispatch(
    stationDefinitionSlice.endpoints.getStationGroupsByNames.initiate(stationGroupQueryProps)
  );

  const stationGroupsQuery = stationDefinitionSlice.endpoints.getStationGroupsByNames.select(
    stationGroupQueryProps
  )(getState());

  const stationsVisibility = {};
  const stationNames = sortBy(
    uniq(
      flatMap(
        stationGroupsQuery?.data?.map(x =>
          x.name === stationGroup.name ? x.stations.map(y => y.name) : []
        )
      )
    )
  );

  const newStationsVisibility = produce(stationsVisibility, draft =>
    stationNames.forEach((stationName: string) => {
      draft[stationName] = {
        visibility: true,
        isStationExpanded: false,
        stationName
      };
    })
  );

  dispatch(waveformActions.setStationsVisibility(newStationsVisibility));
};

/**
 * Pans in the direction and updates the zoomInterval and viewableInterval accordingly
 *
 * @param panDirection the direction to pan
 * @param options containing @param shouldLoadAdditionalData which determines if it should
 * load new data if panning would take the user out of the currently loaded data (defaults to true),
 * and @param onPanningBoundaryReached which is a callback that is called if a boundary is reached.
 */
export const pan = (
  panDirection: WaveformTypes.PanType,
  {
    shouldLoadAdditionalData = true,
    onPanningBoundaryReached
  }: { shouldLoadAdditionalData?: boolean; onPanningBoundaryReached?: () => void }
) => (dispatch: AppDispatch, getState: () => AppState): CommonTypes.TimeRange => {
  const { waveform } = getState().app;
  const { viewableInterval, zoomInterval } = waveform;
  const { timeRange: currentInterval } = getState().app.workflow;
  const processingAnalystConfigurationQuery = processingConfigurationApiSlice.endpoints.getProcessingAnalystConfiguration.select()(
    getState()
  );

  let pannedViewTimeInterval = Utils.getPannedTimeInterval(
    panDirection,
    zoomInterval,
    processingAnalystConfigurationQuery.data.waveformPanRatio
  );

  // if panning boundary is reached, will adjust start/end time and invoke callback
  pannedViewTimeInterval = Utils.adjustIntervalIfBoundaryReached(
    pannedViewTimeInterval,
    currentInterval,
    processingAnalystConfigurationQuery.data.waveformPanningBoundaryDuration,
    shouldLoadAdditionalData ? onPanningBoundaryReached : undefined
  );

  // if not loading new data, if we reach the edge of the viewable interval, stop.
  if (!shouldLoadAdditionalData) {
    pannedViewTimeInterval = Utils.adjustIntervalIfBoundaryReached(
      pannedViewTimeInterval,
      viewableInterval,
      0
    );
  }

  const newViewDuration = pannedViewTimeInterval.endTimeSecs - pannedViewTimeInterval.startTimeSecs;
  const newZoomInterval: CommonTypes.TimeRange =
    panDirection === WaveformTypes.PanType.Left
      ? {
          startTimeSecs: pannedViewTimeInterval.startTimeSecs,
          endTimeSecs: pannedViewTimeInterval.startTimeSecs + newViewDuration
        }
      : {
          startTimeSecs: pannedViewTimeInterval.endTimeSecs - newViewDuration,
          endTimeSecs: pannedViewTimeInterval.endTimeSecs
        };

  const possibleRangeOfDataToLoad = Utils.getPossibleDataRangeInterval(
    pannedViewTimeInterval,
    currentInterval,
    viewableInterval,
    panDirection,
    processingAnalystConfigurationQuery.data.minimumRequestDuration,
    processingAnalystConfigurationQuery.data.waveformPanningBoundaryDuration
  );

  // determine if we need to load data or just pan the current view
  // floor/ceil the values to minimize the chance of erroneous reloading
  if (
    Math.ceil(possibleRangeOfDataToLoad.startTimeSecs) <
      Math.floor(viewableInterval.startTimeSecs) ||
    Math.floor(possibleRangeOfDataToLoad.endTimeSecs) > Math.ceil(viewableInterval.endTimeSecs)
  ) {
    dispatch(
      waveformActions.setViewableInterval({
        startTimeSecs: Math.min(
          viewableInterval.startTimeSecs,
          possibleRangeOfDataToLoad.startTimeSecs
        ),
        endTimeSecs: Math.max(viewableInterval.endTimeSecs, possibleRangeOfDataToLoad.endTimeSecs)
      })
    );
  }
  return newZoomInterval;
};

/**
 * Sets the station visibility for the provided station. Will create a new StationVisibilityChanges object
 * if none exists in the existing store.
 *
 * @param station a station or station name for which to set the visibility
 * @param isVisible whether the station should be visible or not.
 */
export const setStationVisibility = (
  station: StationTypes.Station | string,
  isVisible: boolean
) => (dispatch: AppDispatch, getState: () => AppState): void => {
  const { stationsVisibility } = getState().app.waveform;
  const newVisMap = produce(stationsVisibility, draft => {
    const stationName = Utils.getStationName(station);
    draft[stationName] = draft[stationName] ?? Utils.newStationVisibilityChangesObject(stationName);
    draft[stationName].visibility = isVisible;
  });
  dispatch(waveformActions.setStationsVisibility(newVisMap));
};

/**
 * Sets the station to be expanded. Will create a new StationVisibilityChanges object
 * if none exists in the existing store.
 *
 * @param station a station or station name for which to set the visibility
 * @param isVisible whether the station should be visible or not.
 */
export const setStationExpanded = (station: StationTypes.Station | string, isExpanded = true) => (
  dispatch: AppDispatch,
  getState: () => AppState
): void => {
  const { stationsVisibility } = getState().app.waveform;
  const newVisMap = produce(stationsVisibility, draft => {
    const stationName = Utils.getStationName(station);
    draft[stationName] = draft[stationName] ?? Utils.newStationVisibilityChangesObject(stationName);
    draft[stationName].isStationExpanded = isExpanded;
  });
  dispatch(waveformActions.setStationsVisibility(newVisMap));
};

/**
 * Sets the channel visibility within the provided station. Will create a new StationVisibilityChanges object
 * if none exists in the existing store.
 *
 * @param station a station or station name for which to set the channel's visibility.
 * @param channel a channel or channel name for which to set the visibility.
 * @param isVisible whether the station should be visible or not.
 */
export const setChannelVisibility = (
  station: StationTypes.Station | string,
  channelName: ChannelTypes.Channel | string,
  isVisible: boolean
) => (dispatch: AppDispatch, getState: () => AppState): void => {
  const { stationsVisibility } = getState().app.waveform;
  const newVisMap = produce(stationsVisibility, draft => {
    const vis: StationVisibilityChanges =
      draft[Utils.getStationName(station)] ??
      Utils.newStationVisibilityChangesObject(Utils.getStationName(station));
    if (isVisible) {
      draft[Utils.getStationName(station)] = Utils.getChangesForVisibleChannel(vis, channelName);
    } else {
      draft[Utils.getStationName(station)] = Utils.getChangesForHiddenChannel(vis, channelName);
    }
  });

  dispatch(waveformActions.setStationsVisibility(newVisMap));
};

/**
 * Sets all channels within a station to visible.
 *
 * @param station the station or station name for which to show all channels
 */
export const showAllChannels = (station: StationTypes.Station | string) => (
  dispatch: AppDispatch,
  getState: () => AppState
): void => {
  const { stationsVisibility } = getState().app.waveform;
  const newVisMap = produce(stationsVisibility, draft => {
    draft[Utils.getStationName(station)].hiddenChannels = undefined;
  });
  dispatch(waveformActions.setStationsVisibility(newVisMap));
};

export const Operations = {
  initializeStationVisibility,
  initializeWaveformIntervals,
  pan,
  resetStationsVisibility,
  setStationsVisibility,
  setStationVisibility,
  setStationExpanded,
  setChannelVisibility,
  setZoomInterval,
  showAllChannels
};
