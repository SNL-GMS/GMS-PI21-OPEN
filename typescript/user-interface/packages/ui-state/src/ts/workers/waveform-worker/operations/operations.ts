/** Web worker operations */
export const WorkerOperations = {
  FETCH_SIGNAL_DETECTIONS_WITH_SEGMENTS_BY_STATIONS_TIME:
    'fetchSignalDetectionsWithSegmentsByStationsAndTime',
  FETCH_CHANNEL_SEGMENTS_BY_CHANNEL: 'fetchChannelSegmentsByChannel',
  FETCH_EVENTS_WITH_DETECTIONS_AND_SEGMENTS_BY_TIME: 'fetchEventsWithDetectionsAndSegmentsByTime',
  GET_WAVEFORM: 'getWaveform',
  GET_BOUNDARIES: 'getBoundaries',
  CLEAR_WAVEFORMS: 'clearWaveforms',
  CANCEL_WORKER_REQUESTS: 'cancelWorkerRequests'
};
