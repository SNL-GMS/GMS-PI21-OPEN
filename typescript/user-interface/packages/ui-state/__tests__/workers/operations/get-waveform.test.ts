/* eslint-disable @typescript-eslint/no-magic-numbers */

import type { WeavessTypes } from '@gms/weavess-core';
import Axios from 'axios';

import type { GetWaveformParams } from '../../../src/ts/workers/waveform-worker/operations/get-waveform';
import { getWaveform } from '../../../src/ts/workers/waveform-worker/operations/get-waveform';
import { WaveformStore } from '../../../src/ts/workers/waveform-worker/worker-store/waveform-store';

const timeRange: WeavessTypes.TimeRange = {
  startTimeSecs: 0,
  endTimeSecs: 3
};
const params: GetWaveformParams = {
  id: 'TestId',
  startTime: timeRange.startTimeSecs,
  endTime: timeRange.endTimeSecs,
  domainTimeRange: timeRange
};

const positionBuffer = Float32Array.from([0, 100, 25, 102, 50, 104, 75, 105, 100, 130]);

describe('getWaveform', () => {
  it('has a WaveformStore it can access', () => {
    expect(WaveformStore).toBeDefined();
  });

  it('Can get boundaries that were previously set', async () => {
    const wavePromise = new Promise<Float32Array>(resolve => {
      resolve(positionBuffer);
    });
    await WaveformStore.store(params.id, wavePromise);
    expect(WaveformStore.retrieve(params.id)).toBeDefined();
    const result = await getWaveform(params);
    expect(result).toEqual(positionBuffer);
  });

  it('get Position Buffer with bad time range', async () => {
    const badParams = {
      ...params,
      startTime: 1,
      endTime: 0
    };
    const wavePromise = new Promise<Float32Array>(resolve => {
      resolve(positionBuffer);
    });
    await WaveformStore.store(badParams.id, wavePromise);
    expect(WaveformStore.retrieve(badParams.id)).toBeDefined();
    Axios.request = jest
      .fn()
      .mockRejectedValue(
        new Error(
          'Start index should never be greater than end index. Something is wrong with the logic.'
        )
      );
    await expect(getWaveform(badParams)).rejects.toThrow(
      'Start index should never be greater than end index. Something is wrong with the logic.'
    );
  });

  it('get Position Buffer should return partial buffer', async () => {
    const partialParams = {
      ...params,
      startTime: 0,
      endTime: 1
    };
    const wavePromise = new Promise<Float32Array>(resolve => {
      resolve(positionBuffer);
    });
    await WaveformStore.store(partialParams.id, wavePromise);
    expect(WaveformStore.retrieve(partialParams.id)).toBeDefined();
    const result = await getWaveform(partialParams);
    expect(result).toMatchSnapshot();
  });

  it('get Position Buffer should return empty buffer', async () => {
    const emptyParams = {
      ...params,
      startTime: 4,
      endTime: 5
    };
    const wavePromise = new Promise<Float32Array>(resolve => {
      resolve(positionBuffer);
    });
    await WaveformStore.store(emptyParams.id, wavePromise);
    expect(WaveformStore.retrieve(emptyParams.id)).toBeDefined();
    const result = await getWaveform(emptyParams);
    expect(result).toMatchSnapshot();
  });
});
