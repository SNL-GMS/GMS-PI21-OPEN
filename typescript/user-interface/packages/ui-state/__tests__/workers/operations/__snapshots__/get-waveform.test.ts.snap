// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getWaveform get Position Buffer should return empty buffer 1`] = `Float32Array []`;

exports[`getWaveform get Position Buffer should return partial buffer 1`] = `
Float32Array [
  0,
  100,
  25,
  102,
]
`;
